zilliqa/src/block_store.rs

```diff
@@ -1,25 +1,333 @@
+        Self {
+            parent_hash,
+            from,
+            proposal,
+        }
+    }
+}
+
+/// A block cache.
+/// We need to be careful to conserve block space in the presence of block flooding attacks, and we need to
+/// make sure we don't lose blocks that form part of the main chain repeatedly, else we will never be able
+/// to construct it.
+/// Similarly, we should ensure that we always buffer proposals close to the head of the tree, else we will
+/// lose sync frequently and have to request, which will slow down block production.
+///
+/// An easy way to do this is to put a hash of the node address (actually, we just use the low bits) in the
+/// bottom (log2(N_WAYS)) bits of the view number. We then evict the largest tag le (max_view - buffer).
+///
+/// I don't think it actually matters whether we use the view or the block number here, since we're not using
+/// fixed-size arrays.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct BlockCache {
+    /// Caches proposals that are not yet blocks, and are before the tail cache.
+    pub cache: BTreeMap<u128, BlockCacheEntry>,
+    /// Caches proposals close to the head.
+    pub tail: BTreeMap<u128, BlockCacheEntry>,
+    /// Caches ranges where we think there is no block at all (just an empty view)
+    pub empty_view_ranges: RangeMap,
+    /// The head cache - this caches
+    /// An index into the cache by parent hash
+    pub by_parent_hash: HashMap<Hash, HashSet<u128>>,
+    /// Set associative shift
+    pub shift: usize,
+    /// This is used to count the number of times we've looked for a fork.
+    /// The counter is zeroed when we receive (or pop) a new block, and counts 1 every
+    /// time we looked.
+    pub fork_counter: usize,
+    /// Copied from the parent to minimise the number of additional parameters we need.
+    pub max_blocks_in_flight: u64,
+}
+
+impl BlockCache {
+    pub fn new(max_blocks_in_flight: u64) -> Self {
+        Self {
+            cache: BTreeMap::new(),
+            tail: BTreeMap::new(),
+            empty_view_ranges: RangeMap::new(),
+            by_parent_hash: HashMap::new(),
+            shift: 8 - constants::BLOCK_CACHE_LOG2_WAYS,
+            fork_counter: 0,
+            max_blocks_in_flight,
+        }
+    }
+
+    pub fn key_from_view(&self, peer: &PeerId, view_num: u64) -> u128 {
+        let ways = peer.to_bytes().pop().unwrap_or(0x00);
+        u128::from(ways >> self.shift) | (u128::from(view_num) << self.shift)
+    }
+
+    pub fn view_from_key(&self, key: u128) -> u64 {
+        u64::try_from(key >> self.shift).unwrap()
+    }
+
+    pub fn min_key_for_view(&self, view: u64) -> u128 {
+        u128::from(view) << self.shift
+    }
+
+    pub fn min_tail_key(&self, highest_known_view: u64) -> u128 {
+        let highest_key = u128::from(highest_known_view + 1) << self.shift;
+        highest_key - u128::try_from(constants::BLOCK_CACHE_TAIL_BUFFER_ENTRIES).unwrap()
```

Comment:
What is the purpose of taking `BLOCK_CACHE_TAIL_BUFFER_ENTRIES` off this value?

Reply:
This function returns the minimum key ( view_number << log_2(ways) ) which will be retained in the tail cache.
highest_key is the highest key we think might ever exist in the system.
One subtracts the number of entries in the tail cache from the highest key to find the key above which we are prepared to store entries in the tail cache.
Subtract too much and the tail cache will become larger than you want. Subtract too little and it will become smaller - subtract nothing and the only entries you will be prepared to store are those that don't exist yet.

Reply:
This function returns the minimum key ( view_number << log_2(ways) ) which will be retained in the tail cache.
highest_key is the highest key we think might ever exist in the system.
One subtracts the number of entries in the tail cache from the highest key to find the key above which we are prepared to store entries in the tail cache.
Subtract too much and the tail cache will become larger than you want. Subtract too little and it will become smaller - subtract nothing and the only entries you will be prepared to store are those that don't exist yet.

(I'll add a comment in the code :-)  - thanks for pointing out!)

Reply:
This function returns the minimum key ( view_number << log_2(ways) ) which will be retained in the tail cache.
highest_key is the highest key we think might ever exist in the system.
One subtracts the number of entries in the tail cache from the highest key to find the key above which we are prepared to store entries in the tail cache.
Subtract too much and the tail cache will become larger than you want. Subtract too little and it will become smaller - subtract nothing and the only entries you will be prepared to store are those that don't exist yet.

(I'll add a comment in the code :-)  - thanks for pointing out!)

Reply:
Thanks! Understood. Deriving this from `self.highest_known_view` (which is set during in `buffer_proposal` just before we insert), means that the `tail` will only contain proposals that are >= the largest view we've seen. So if we are block 100 and syncing, and we receive new block 100_000, we put only proposals with view >= 100_000 in `tail`. 

So this acts as a mechanism to buffer new activity happening at the head whilst connected to the network but syncing. The size of this cache should then depend on sync times - if in sync time there are 1000 new proposals at the head then this cache doesnt do much.

Please correct me if that isnt accurate! I'll propose changing the name to `look_ahead_cache` or something in a follow up PR

Reply:
Yes - it's intended to cure a problem where the head moves on as you are fetching the blocks up to it. Two things then happen - first, as you say, you are then throwing away broadcast proposals so as to fetch them later, which is wasteful (but the main cache ought to catch them if you really wanted them[1]), and second, by the time you've received a block as a result of one of your requests, the head has moved on and so there's another block to fetch .. 

As such, the only capacity you really need here is one RTT (maybe two)  - at one block per second 16 entries gives you 15 or so seconds, which I agree may not be enough, but I thought it was!

If you're interested in buffering blocks generally, the main block cache should do that for you, advancing its lower bound as you add blocks to your chain.

I'd go for `head_cache` or similar (`look_ahead_cache` kinda implies we're looking ahead of the chain, which we're not - though that is another way to solve the problem: remove this cache and just have the main cache, and simply request views ahead of where you are in the belief that they probably exist).

[1] I took the view that you shouldn't attempt to fix this problem too hard in the tail cache, because you only have a limited amount of memory space, and if you're after enough space to buffer the gap between the currently broadcast block and the last synced block, you should allocate it to the main buffer cache.